src/coin-selection/lib/Convex/Query.hs

```diff
@@ -97,72 +91,74 @@ instance MonadIO m => MonadUtxoQuery (WalletAPIQueryT m) where
-  (MonadBlockchain m, MonadUtxoQuery m, MonadError (BalanceAndSubmitError C.ConwayEra) m) =>
+  (MonadBlockchain era m, MonadUtxoQuery era m, MonadError (BalanceAndSubmitError era) m, C.IsBabbageBasedEra era) =>
   Tracer m TxBalancingMessage ->
   C.PaymentCredential ->
-  Maybe (C.TxOut C.CtxTx C.ConwayEra) ->
-  TxBuilder ->
+  Maybe (C.TxOut C.CtxTx era) ->
+  TxBuilder era ->
   ChangeOutputPosition ->
-  m (C.Tx C.ConwayEra)
+  m (C.Tx era)
 balancePaymentCredential dbg cred = balancePaymentCredentials dbg cred []
 
 -- | Balance a transaction body, sign it with the operator's key, and submit it to the network.
 balanceAndSubmitOperator ::
-  (MonadBlockchain m, MonadUtxoQuery m, MonadError (BalanceAndSubmitError C.ConwayEra) m) =>
+  forall era m.
+  (MonadBlockchain era m, MonadUtxoQuery era m, MonadError (BalanceAndSubmitError era) m, C.IsBabbageBasedEra era) =>
   Tracer m TxBalancingMessage ->
   Operator Signing ->
-  Maybe (C.TxOut C.CtxTx C.ConwayEra) ->
-  TxBuilder ->
+  Maybe (C.TxOut C.CtxTx era) ->
+  TxBuilder era ->
   ChangeOutputPosition ->
-  m (C.Tx C.ConwayEra)
+  m (C.Tx era)
 balanceAndSubmitOperator dbg op changeOut txBody changePosition =
   balancePaymentCredential dbg (operatorPaymentCredential op) changeOut txBody changePosition
-    >>= signAndSubmitOperator op
+    >>= inBabbage @era signAndSubmitOperator op
 
 {-| Add the operator's signature to the transaction and send it to the blockchain
 -}
 signAndSubmitOperator
-  :: (MonadBlockchain m, MonadError (BalanceAndSubmitError C.ConwayEra) m)
+  :: (MonadBlockchain era m, MonadError (BalanceAndSubmitError era) m, C.IsShelleyBasedEra era)
   => Operator Signing
-  -> C.Tx C.ConwayEra
-  -> m (C.Tx C.ConwayEra)
+  -> C.Tx era
+  -> m (C.Tx era)
 signAndSubmitOperator op tx = do
   let finalTx = signTxOperator op tx
-  liftResult SubmitError (sendTx finalTx) $> finalTx
+  mapError SubmitError (sendTx finalTx) $> finalTx
 
 {-| UTxOs that are locked by the operator's payment credential
 |-}
-operatorUtxos :: MonadUtxoQuery m => Operator k -> m [(C.TxIn, C.InAnyCardanoEra (C.TxOut C.CtxUTxO))]
+operatorUtxos :: MonadUtxoQuery era m => Operator k -> m [(C.TxIn, C.TxOut C.CtxUTxO era)]
 operatorUtxos = fmap (Map.toList . fmap fst . _utxos) . utxosByPaymentCredential . operatorPaymentCredential
 
 {-| Select a single UTxO that is controlled by the operator. |-}
-selectOperatorUTxO :: MonadUtxoQuery m => Operator k -> m (Maybe (C.TxIn, C.InAnyCardanoEra (C.TxOut C.CtxUTxO)))
+selectOperatorUTxO :: MonadUtxoQuery era m => Operator k -> m (Maybe (C.TxIn, C.TxOut C.CtxUTxO era))
 selectOperatorUTxO operator = fmap listToMaybe (operatorUtxos operator)
 
 -- | Failures during txn balancing and submission
 data BalanceAndSubmitError era =
   BalanceError (BalanceTxError era)
-  | SubmitError String
-  deriving stock (Generic, Show)
-  deriving anyclass (ToJSON, FromJSON)
+  | SubmitError (ValidationError era)
+  deriving stock Show
+  deriving anyclass Exception
```

Comment:
We keep the original `ValidationError era` instead of the string. We could still convert to JSON after `show`ing the `BalanceAndSubmitError`, but now we can get some more information from matching on the constructor of `ValidationError`